components/server/src/workspace/workspace-starter.ts

```diff
@@ -684,127 +684,166 @@ export class WorkspaceStarter {
-            // few bytes of JSON in the database for no good reason.
-            configuration.featureFlags = featureFlags;
-        }
+            if (!!featureFlags) {
+                // only set feature flags if there actually are any. Otherwise we waste the
+                // few bytes of JSON in the database for no good reason.
+                configuration.featureFlags = featureFlags;
+            }
 
-        const usageAttributionId = await this.userService.getWorkspaceUsageAttributionId(user, workspace.projectId);
-
-        const now = new Date().toISOString();
-        const instance: WorkspaceInstance = {
-            id: uuidv4(),
-            workspaceId: workspace.id,
-            creationTime: now,
-            ideUrl: "", // Initially empty, filled during starting process
-            region: this.config.installationShortname, // Shortname set to bridge can cleanup workspaces stuck preparing
-            workspaceImage: "", // Initially empty, filled during starting process
-            status: {
-                version: 0,
-                conditions: {},
-                phase: "preparing",
-            },
-            configuration,
-            usageAttributionId,
-        };
-        if (WithReferrerContext.is(workspace.context)) {
-            this.analytics.track({
-                userId: user.id,
-                event: "ide_referrer",
-                properties: {
-                    workspaceId: workspace.id,
-                    instanceId: instance.id,
-                    referrer: workspace.context.referrer,
-                    referrerIde: workspace.context.referrerIde,
-                },
+            const usageAttributionId = await this.userService.getWorkspaceUsageAttributionId(user, workspace.projectId);
+
+            let workspaceClass = "";
+            let classesEnabled = await getExperimentsClientForBackend().getValueAsync("workspace_classes", false, {
+                user: user,
             });
+
+            if (classesEnabled) {
+                // this is either the first time we start the workspace or the workspace was started
+                // before workspace classes and does not have a class yet
+                if (!previousInstance?.workspaceClass) {
+                    if (workspace.type == "regular") {
+                        if (user.additionalData?.workspaceClasses?.regular) {
+                            workspaceClass = user.additionalData?.workspaceClasses?.regular;
+                        }
+                    }
+
+                    if (workspace.type == "prebuild") {
+                        if (user.additionalData?.workspaceClasses?.prebuild) {
+                            workspaceClass = user.additionalData?.workspaceClasses?.prebuild;
+                        }
+                    }
+
+                    if (!workspaceClass) {
+                        workspaceClass = this.config.workspaceClasses.find((cl) => cl.isDefault)?.id ?? "";
```

Comment:
I feel this should be in a function inside a namesapce `WorkspaceClasses` so that we can test it etc.
@Furisto Let me find a couple of minutes and I can provide a commit to show what I mean.

Reply:
@Furisto Here is a commit showing what I mean: https://github.com/gitpod-io/gitpod/commit/7412eb0f4ebc99f71a08bcd49a68f2c7f8a4e72f

I think it make sense to incorporate that commit + [exposing of that config under "experimental"](https://github.com/gitpod-io/gitpod/pull/11313/files#r924324654) into this PR, and then we should be fine to merge. :heavy_check_mark: 

:cloud: Ideally we should also have tests for those functions.